Math / tuple library improvements

tdhsmith opened the pull request:
This is primarily a cleanup/reimagining of my work from 2018 that never got merged, but it's got a number of new bits too.

- fused modular exponentiation `_^_mod_`, providing much much faster operation than composition of the unfused method
- strong probable primality test (b-SPRP) `_is a strong probable prime to base_`
- deterministic primality tests `_is prime` via
  - look up table for small primes in `[0..130]`
  - Forǐsek & Jančina's _FJ32_256_ algorithm for `[0..2^32)`
  - Deterministic SPRP using Sinclair's bases for `[0..2^64]`
- small semantic restriction improvements, including fix for a major bug in tuple element removal `_-_`
- migrate names for floor and ceiling methods from box drawing characters to appropriate Unicode glyphs for floor and ceiling
- exact integer square root method `⌊√_⌋`
- extend tuple sum and product methods `∑_` and `∏_` to work on all number tuples (previously only integer tuples)
- add multiple-split method `explode_at element_«skipping empties»?` for breaking a tuple into sub-tuples by a delimiter element
- add `drain_` macro for simple reading of an entire stream to tuple

tdhsmith commented:
This could certainly use some tests before merge, however I'm a little unsure on testing methodology for things like the primality check -- short lists of input and output seem prone to missing smaller bugs, but more comprehensive tests are either slow or non-deterministic.

Is the general premise that the test suite itself is primarily for detecting regression issues and we've _separately_ audited that the calculation is correct?

toddATavail commented on distro/src/avail/Avail.avail/Foundation.avail/Tuples.avail:
```diff
@@ -3243,18 +3344,36 @@ Semantic restriction "_'s⁇last" is
 Public stable method "∑_" is
 [
-	intTuple : extended integer*
+	intTuple : extended integer+
```
This should probably continue to work for empty tuples, using zero as the natural base accumulator value.

toddATavail commented on distro/src/avail/Avail.avail/Foundation.avail/Tuples.avail:
```diff
@@ -3243,18 +3344,36 @@ Semantic restriction "_'s⁇last" is
+	intTuple : extended integer+
 |
-	0
+	left fold intTuple through
+		[i : extended integer, ri : extended integer | i + ri]
```
To support an empty tuple, you'll want to augment the operation tuple:

```avail
left fold <0> ++ intTuple through
		[i : extended integer, ri : extended integer | i + ri]
```

toddATavail commented on distro/src/avail/Avail.avail/Foundation.avail/Tuples.avail:
```diff
@@ -3243,18 +3344,36 @@ Semantic restriction "_'s⁇last" is
+ */
+Public stable method "∑_" is
+[
+	numTuple : number+
```
You'll want to make this work for empty tuples also, in a similar way to the way that I described above.

toddATavail commented on distro/src/avail/Avail.avail/Foundation.avail/Tuples.avail:
```diff
@@ -3333,27 +3452,78 @@ Semantic restriction "∑_" is
+ */
+Semantic restriction "∑_" is
+	a closed element operation on nonempty tuples of double;
+/**
```
Add a blank line here.

toddATavail commented on distro/src/avail/Avail.avail/Foundation.avail/Tuples.avail:
```diff
@@ -3333,27 +3452,78 @@ Semantic restriction "∑_" is
 
 /**
- * Take the product of the elements of a tuple of integers.
+ * Non-empty tuple of doubles will return double
```
Add a terminal period.

toddATavail commented on distro/src/avail/Avail.avail/Foundation.avail/Tuples.avail:
```diff
@@ -3333,27 +3452,78 @@ Semantic restriction "∑_" is
+[
+	numTuple : number*
+|
+	1
```
Probably a good default, being the multiplicative identity. Good thinking.

toddATavail commented on src/main/kotlin/com/avail/compiler/AvailCompiler.kt:
```diff
@@ -3564,6 +3564,12 @@ class AvailCompiler(
 			onSuccess(solutions) { this.rollbackModuleTransaction(it) }
 		}
 		recordExpectationsRelativeTo(start)
+		if (loader.lexicalScanner().allVisibleLexers.isEmpty())
+		{
+			start.expected(
+				STRONG,
+				"module to export at least 1 lexer in order to handle command")
+		}
```
I would like Mark to review this, to make sure that it's in the right place doing the right thing at the right time. The compiler is really tricky. :)

tdhsmith replied:
Quick note that this inclusion was accidental -- I have another branch for experimenting with workbench tweaks -- I will probably roll it back and wait until I have a PR for that branch since I haven't even analyzed it myself.

toddATavail replied:
Okay, good idea. Roll it back and we'll come back to it when you're ready with your workbench updates. Separately, if you are interested in working on workbench items, maybe you'd like to join us for the Anvil IDE project? It is coming along nicely, and could really benefit from your Avail interest and web expertise. Contact me privately if you're interested. :)

tdhsmith commented on avail-bootstrap/src/main/java/com/avail/tools/bootstrap/generated/PrimitiveNames_en.properties:
```diff
@@ -1277,7 +1277,7 @@ P_ConstructDoubleFromParts_comment=\
 \ */\n
 # DoubleCeiling : _=1
-P_DoubleCeiling=⎡_⎤
+P_DoubleCeiling=⌈_⌉
```
Hrm, that fits an idea I had elsewhere, that it might be worth starting a pattern of "migration modules" for each release version transition.

markATavail commented on distro/src/avail/Avail.avail/Foundation.avail/Late Math.avail:
```diff
@@ -846,3 +868,247 @@ Public stable method "_(decimal)" is
+	73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127
+};
+
+/**
+ * Determine if a {@type "[0..130]" small integer} is prime. This method performs
+ * a lookup against a list of the first 31 primes.
+ *
+ * @method "_is prime"
+ * @param "x" "whole number"
+ * @returns "boolean"
+ *          {@method "true"} if {@param "x"} is prime, {@method "false"}
+ *          otherwise.
+ * @category "Mathematics" "Integers" "Numbers"
+ */
+Public method "_is prime" is
+[
+	x : [0..130]
+|
+	x∈smallPrimes
+];
```
Agreed, but the general case method should be an actual implementation of an exact primality test.  Defining the abstract method and overriding it for [131..∞) seems risky, by repeating the threshold (+1 in this case) in two places.  Divisor testing in a loop should suffice, although a sieve is technically better performance asymptotically.  It's questionable whether it's worthwhile to statistically check for primality first, and answer false if it was determined to be definitely composite (and falling back on the sieve otherwise).  I wouldn't bother with the pre-check.

tdhsmith replied:
Ok I'm happy to put a base implementation in if that's preferred. I shied away from it because the decent algorithms at that scale are complex, but I understand getting the type coverage might be more important than only accepting code with good performance.

> It's questionable whether it's worthwhile to statistically check for primality first

Funny, I see so much advice on the internet that starts with "first throw Miller-Rabin at it" but it's hard to judge whether those posts come from any point of experience. In any case I already have the statistical test, so I'll try it both with and without.

markATavail commented on distro/src/avail/Avail.avail/Foundation.avail/Late Math.avail:
```diff
@@ -846,3 +868,247 @@ Public stable method "_(decimal)" is
+	/* Check small divisors and ensure x is coprime with 210 */
+	If (x is even) ∨ (x mod 3 = 0) ∨ (x mod 5 = 0) ∨ (x mod 7 = 0) then
+	[
+		Exit body with false
```
Please use Exit_with_if_.

markATavail commented on distro/src/avail/Avail.avail/Foundation.avail/Late Math.avail:
```diff
@@ -846,3 +868,247 @@ Public stable method "_(decimal)" is
+ */
+Public method "_is prime" is
+[
+	x : [0..2^64]
```
With an abstract method covering [0..∞), this can simply be an override for [2^32..2^64), making the below cast into k unnecessary.

markATavail commented on distro/src/avail/Avail.avail/Foundation.avail/Tuples.avail:
```diff
@@ -293,6 +294,25 @@ Public method "_'s⁇strengthening function" is
+ * quick semantic restrictions on number tuples returning that tuple's element
+ * type, given that it is uniform.
+ */
+Macro "a closed element operation on nonempty tuples of_" is
```
Very clean macro!

tdhsmith commented on distro/src/avail/Avail.avail/Foundation.avail/Tuples.avail:
```diff
@@ -2517,22 +2537,103 @@ Semantic restriction "split_before_" is
+*          matching delimiter, with that delimiting element removed.
+* @category "Tuples" "Transformers"
+*/
+Public stable method "explode_at element_«skipping empties»?" is
```
In a vacuum I like "exploded" more but since I built it as extension of `split_before_`/`split_after_` I think that biased me. I'll ruminate.

I see an argument for pure transformations of immutable values preferring past-participle postfix names and more stateful, side-effect-laden methods preferring present imperative prefix names, but it probably still comes down to taste most of the time. E.g. `_split before_` doesn't sound as good to me (the present/past/past participle ambiguity of split is at least partially to blame).

tdhsmith commented on distro/src/avail/Avail.avail/Foundation.avail/Late Math.avail:
```diff
@@ -846,3 +868,247 @@ Public stable method "_(decimal)" is
+	exponent : whole number's type,
+	modulus : natural number's type
+|
+	(base ^ exponent) mod modulus
```
Good catch, these restrictions are generally going to err to the "unreasonable" end of the spectrum.

> If the exponent bound is always < 2, intersect the base type with `[0..⌈modulus⌉)`

I don't think intersect is exactly the right operation, since disjoint base and modulus types should still have nonempty result types. (e.g. base `[10..12]` mod `[4..5]`'s has possible values in `[0..3]`).

Stepping back, this method is basically designed for trying to make larger bases and exponents calculable, so I'm not sure that returning anything stricter than `[0..⌈modulus⌉)` is actually going to have real world benefit..? I'll try to sketch the cases out though.

markATavail replied:
True dat!  `"_mod_"` has a pretty good semantic restriction for the case that the exact divisor is known.  It checks if the quotients of the largest dividend (over divisor) and the smallest dividend (over divisor) are equal, and if so, it computes the two remainders and uses the range spanning them.  Otherwise it uses [0..divisor), since both 0 and divisor-1 must both be possible results in that case.

I suppose you could generalize that to a variable base, but I'm sure you're right about diminishing benefits beyond `[0..⌈modulus⌉)`.  Besides, multiple semantic restrictions can make varying claims about the resulting type, and their intersection is used for the expected return type at a call site, so someone could always add a stricter case if they felt like coding it themselves.

tdhsmith commented:
Alright:

- unit tests are in, including some fixes to the stream tests
- `drain_` also covers readers -- since the stream version is oriented towards its output (and not side effects), providing one for writers / general iterators didn't make as much sense to me but I could be persuaded
- added naive `_is prime` implementation for all natural numbers, using trial division
- also added Fermat factorization method because why not, and then exposed a public `factor_` method and `prime factorization of_` macro
- finally decided to stop because scope is a thing

I would like to _one day_ upgrade the base primality test to quadratic sieve, but that would logically fit after having some matrix math support... _whiiiich_ is certainly something I'd like to work on but I'm not sure I want to approach it immediately, or that it should even live in the avail root.

Let me know if it needs rebasing or any other revisions based on the new work!

markATavail commented:
Sorry for the delay.  I didn't realize I was holding it up.
